synapse/storage/databases/main/stream.py

```diff
@@ -1099,6 +1115,7 @@ def _paginate_room_events_txn(
         from_token: RoomStreamToken,
         to_token: Optional[RoomStreamToken] = None,
         direction: str = "b",
+        order_by: str = "topological",
```

Comment:
Instead of adding a new `order_by` argument, I think we can just check whether `from_token.topological is not None` to determine whether to sort by `topological` or `stream_ordering`

Reply:
My concern with this approach is the magic nature of it, a reader new to the code won’t know if a token is topological or not and thus the ordering becomes much harder to reason about.
Unless I’m missing something on the token definition that is! Happy to switch to using the token detection if that would be preferred?

Reply:
I think we just want it to work correctly with whatever token is provided.

Is there a case where we want to sort by `depth`(`topological_ordering`) when we give it a stream token? Or vice versa?

Reply:
Makes sense - I did spot other cases where the token defines this kind of behavior elsewhere as well since my initial change; added https://github.com/matrix-org/synapse/pull/12319/commits/ef07fcda333cd22825661eee5f8615b258ac3505 to address.

Reply:
Hm. This appears to have broken some complement tests now!

Reply:
> Hm. This appears to have broken some complement tests now!

Looking into this more. I think this may be a bug in [`get_messages(...)`](https://github.com/matrix-org/synapse/blob/bebf994ee804ef63ce16801c6694713fcd685320/synapse/handlers/pagination.py#L419). 

The following code will always return a `stream_ordering` token but when we're working with `/messages`, it should be a topological token (I think). With the previous code, I feel like it just masked this problem because we always did the `topological_order` sort with `_paginate_room_events_txn` before regardless of what the token was.

https://github.com/matrix-org/synapse/blob/bebf994ee804ef63ce16801c6694713fcd685320/synapse/handlers/pagination.py#L444

Relevant call stack:

```
on_GET (/messages)
get_messages
paginate_room_events
_paginate_room_events_txn
```


## Debugging that failing Complement test

Relevant Complement test: https://github.com/matrix-org/complement/blob/1f527b05ea90afb5ab4d164980f8522576d2b219/tests/msc2716_test.go#L84-L187

Running the test:
```
COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh TestImportHistoricalMessages/parallel/Historical_events_resolve_in_the_correct_order
```


See these steps to access database for homeserver after Complement test runs: https://github.com/matrix-org/complement/pull/165/files

```sh
$ docker exec -it complement_fed_1_fed.hs_with_application_service.hs1_1 /bin/bash
$ apt-get update && apt-get install -y sqlite3
$ sqlite3
> .open /conf/homeserver.db
# To get the room ID, just look at any of the events
> SELECT * FROM events;

# Pretty print the output better
> .mode column
> .headers on
```

This is expected `topological_order` sort and how it behaves when the test passes (you can see how this [matches up to the test here](https://github.com/matrix-org/complement/blob/1f527b05ea90afb5ab4d164980f8522576d2b219/tests/msc2716_test.go#L155-L166)):

```
SELECT depth, stream_ordering, type, state_key, event_id FROM events
WHERE events.room_id ='!FEvmZPRLBIEafdrJWQ:hs1'
ORDER BY depth DESC, stream_ordering DESC;
depth  stream_ordering  type                          state_key             event_id
-----  ---------------  ----------------------------  --------------------  --------------------------------------------
# eventIDsBefore
11     12               m.room.message                                      $Wt7wI7MnraTqyDIu83-zWTa6SKOhw2mj3vU9bumrpwg
10     11               m.room.message                                      $NqSxI26ke8xr9B2eCH05gTSDzpzryw1Jde1EuElONwA
# baseInsertionEventID0
10     -3               org.matrix.msc2716.insertion                        $vGkND0_QMEhU12MJkxc7QwNPv5OFI8bfmzg1TPSgrC8
# historical batch 0
10     -4               org.matrix.msc2716.batch                            $E407MKifUI5x_KQMS1JvJ3WAbRPaDRAR814b0IJ0DIg
10     -5               m.room.message                                      $TS2vmOnEL8bQvbNOzKEYLgvsBIaNW_h8RFckjQsnej4
10     -6               m.room.message                                      $wlBalhDWofWEwxz1ZQoEOMilN6veASLHmIL2CReDoLw
10     -7               m.room.message                                      $840NKtUjST7gNG5zteq17-jfGYHkPH2n_hATD3OOH4I
10     -8               org.matrix.msc2716.insertion                        $ww-VPKdpa_vOE5l2GLpKZNBtgbvn2oEKi1c8KlIvLsU
# historical batch 1
10     -10              org.matrix.msc2716.batch                            $v_sDaykItrVQW-HEd3dt7voIoazLpuw9EtN4syyjCjQ
10     -11              m.room.message                                      $G8KlqUhMsknRszhw3dEJznLWqzWf7bKMRLvtVaOd5mU
10     -12              m.room.message                                      $9IoiumsFGnrXIbcnDG_LjIJjqsQU_z35jTx7fxB13iQ
10     -13              m.room.message                                      $3fxfwV6oz9DZ20pwlhWJgExd9n7hsP7k9w3GM1LyQgw
10     -14              org.matrix.msc2716.insertion                        $lK21br0ZyavuewyBlPAr57WZHItKHvG37ffIBeBDgZ4
# eventIDsAfter
9      10               m.room.message                                      $Z-D4EQEzgrwhr-R1_-JjmAUHIGUjusQp5yh8Dbe4Q6Q
8      9                m.room.message                                      $tjnuGZaismCDTPUoHJ87agAaH9CKyTIIfCYsjKRBiSs
# start of room
7      8                m.room.member                 @alice:hs1            $t1JOLDZLR48r91UmWn0kPmRgwt97usvz1p1Zcx7SJqs
6      7                m.room.name                                         $UhxaGndH86gNZTya1Nuv09eBGtMGc2KBvWYhpeSys1k
5      6                m.room.history_visibility                           $l0f9onBohOxAbGDQ8gudGHHpEgBWAZwSbSkdGdjDo7A
4      5                m.room.join_rules                                   $UZwW_RMaYUyyk8x3WbNowoRNhd-lW85QWD6qBdSEly8
3      4                m.room.power_levels                                 $Yooqdyfxw9m-xWzMnkUIcDMlJlDJQjnDDd2CAdF-kzg
2      3                m.room.member                 @the-bridge-user:hs1  $QynAfgtUHeBUjN1YXvQ6Ye5E1hyE9q59bO2TH89HtQg
1      2                m.room.create                                       $gkDWQFo9jfu_w9pbzSypK_dF9aqnw3aCecnG5wQiCGQ
1      -2               m.room.member                 @maria:hs1            $dWDbmnkE_RpnCmD1Q6ZJHCVMDQvWeJ0kSzCc0GBqHSA
1      -9               m.room.member                 @maria:hs1            $hhIfpjLltBmvMjwTALIGSAfIeY-d3n2vwRXG2s-b-QQ
```

This is the `stream_ordering` sort it tries to use which is not the order we expect in the test.
```
SELECT depth, stream_ordering, type, state_key, event_id FROM events
WHERE events.room_id ='!FEvmZPRLBIEafdrJWQ:hs1'
ORDER BY stream_ordering DESC, depth DESC;
depth  stream_ordering  type                          state_key             event_id
-----  ---------------  ----------------------------  --------------------  --------------------------------------------
11     12               m.room.message                                      $Wt7wI7MnraTqyDIu83-zWTa6SKOhw2mj3vU9bumrpwg
10     11               m.room.message                                      $NqSxI26ke8xr9B2eCH05gTSDzpzryw1Jde1EuElONwA
9      10               m.room.message                                      $Z-D4EQEzgrwhr-R1_-JjmAUHIGUjusQp5yh8Dbe4Q6Q
8      9                m.room.message                                      $tjnuGZaismCDTPUoHJ87agAaH9CKyTIIfCYsjKRBiSs
7      8                m.room.member                 @alice:hs1            $t1JOLDZLR48r91UmWn0kPmRgwt97usvz1p1Zcx7SJqs
6      7                m.room.name                                         $UhxaGndH86gNZTya1Nuv09eBGtMGc2KBvWYhpeSys1k
5      6                m.room.history_visibility                           $l0f9onBohOxAbGDQ8gudGHHpEgBWAZwSbSkdGdjDo7A
4      5                m.room.join_rules                                   $UZwW_RMaYUyyk8x3WbNowoRNhd-lW85QWD6qBdSEly8
3      4                m.room.power_levels                                 $Yooqdyfxw9m-xWzMnkUIcDMlJlDJQjnDDd2CAdF-kzg
2      3                m.room.member                 @the-bridge-user:hs1  $QynAfgtUHeBUjN1YXvQ6Ye5E1hyE9q59bO2TH89HtQg
1      2                m.room.create                                       $gkDWQFo9jfu_w9pbzSypK_dF9aqnw3aCecnG5wQiCGQ
1      -2               m.room.member                 @maria:hs1            $dWDbmnkE_RpnCmD1Q6ZJHCVMDQvWeJ0kSzCc0GBqHSA
10     -3               org.matrix.msc2716.insertion                        $vGkND0_QMEhU12MJkxc7QwNPv5OFI8bfmzg1TPSgrC8
10     -4               org.matrix.msc2716.batch                            $E407MKifUI5x_KQMS1JvJ3WAbRPaDRAR814b0IJ0DIg
10     -5               m.room.message                                      $TS2vmOnEL8bQvbNOzKEYLgvsBIaNW_h8RFckjQsnej4
10     -6               m.room.message                                      $wlBalhDWofWEwxz1ZQoEOMilN6veASLHmIL2CReDoLw
10     -7               m.room.message                                      $840NKtUjST7gNG5zteq17-jfGYHkPH2n_hATD3OOH4I
10     -8               org.matrix.msc2716.insertion                        $ww-VPKdpa_vOE5l2GLpKZNBtgbvn2oEKi1c8KlIvLsU
1      -9               m.room.member                 @maria:hs1            $hhIfpjLltBmvMjwTALIGSAfIeY-d3n2vwRXG2s-b-QQ
10     -10              org.matrix.msc2716.batch                            $v_sDaykItrVQW-HEd3dt7voIoazLpuw9EtN4syyjCjQ
10     -11              m.room.message                                      $G8KlqUhMsknRszhw3dEJznLWqzWf7bKMRLvtVaOd5mU
10     -12              m.room.message                                      $9IoiumsFGnrXIbcnDG_LjIJjqsQU_z35jTx7fxB13iQ
10     -13              m.room.message                                      $3fxfwV6oz9DZ20pwlhWJgExd9n7hsP7k9w3GM1LyQgw
10     -14              org.matrix.msc2716.insertion                        $lK21br0ZyavuewyBlPAr57WZHItKHvG37ffIBeBDgZ4
```

Reply:
@Fizzadar I've created https://github.com/matrix-org/synapse/pull/12370 to fix the `/messages` bug causing the Complement failures you were seeing.